runtime/sema/check_binary_expression.go

```diff
@@ -73,9 +75,42 @@ func (checker *Checker) VisitBinaryExpression(expression *ast.BinaryExpression)
 
 		// Right hand side will always be evaluated
 
-		// Visit the expression, with contextually expected type. Use the expected type
-		// only for inferring wherever possible, but do not check for compatibility.
+		// Visit the expression, with contextually expected type.
+		// Use the expected type only for inferring wherever possible,
+		// but do not check for compatibility.
 		// Compatibility is checked separately for each operand kind.
+		//
+		// If there is no contextually expected type,
+		// then expect the right type to have the type of the left side.
+		// For example, this allows a declaration like this to type-check:
+		//
+		// ```
+		// let x = 1 as UInt8
+		// let y = x + 1
+		// ```
+		//
+		// Also, if there is a contextually expected type,
+		// but the left type is a subtype and more specific (i.e not the same),
+		// then use it instead as the expected type for the right type.
+		// For example, this allows declarations like the following to type-check:
+		//
+		// ```
+		// let x = 1 as UInt8
+		// let y: Integer = x + 1
+		// ```
+		//
+		// ```
+		// let string = "this is a test"
+		// let index = 1 as UInt8
+		// let character = string[index + 1]
+		// ```
+
+		if expectedType == nil ||
+			(IsSubType(leftType, expectedType) && !leftType.Equal(expectedType)) {
```

Comment:
and maybe an additional check of `leftType != expectedType` for an early exit?

Reply:
Subtype and equality testing should be roughly equal in complexity, they are very "similar" (rescursing down into the type structure and comparing).

However, I looked into this and realized that the equality check is needed, as subtyping also consideres equal types subtypes. So what we want here is the proper subtype. I've refactored this in d089fd0939ec20f1dd9f68b99613cab35d33ef1f, so we only have to perform one equality check instead of two.

I also added your exit early suggestion in 9b9f1234a904629bf9d62bd988ca7478aefd1d89: Good observation that two pointer-equal types are never proper subtypes.